rdagent/oai/utils/embedding.py

```diff
@@ -0,0 +1,155 @@
+"""
+Simple embedding utilities for handling token limits and text truncation.
+"""
+
+import re
+from typing import List, Optional
+
+from rdagent.log import rdagent_logger as logger
+
+
+def get_embedding_max_length(model_name: str) -> int:
+    """
+    Get max token length for embedding model, with simple fallback.
+
+    Args:
+        model_name: Name of the embedding model
+
+    Returns:
+        Maximum token length (defaults to 8192)
+    """
+    try:
+        from litellm import get_max_tokens
+
+        max_tokens = get_max_tokens(model_name)
+        if max_tokens and max_tokens > 0:
+            return max_tokens
+    except Exception as e:
+        logger.warning(f"Failed to get max tokens for {model_name}: {e}")
+
+    # Default fallback
+    return 8192
+
+
+def estimate_token_count(text: str) -> int:
+    """
+    Rough estimation of token count for text.
+    Based on conservative heuristic: 1 token ≈ 2 characters (for embedding models).
+
+    Args:
+        text: Input text
+
+    Returns:
+        Estimated token count
+    """
+    # Conservative heuristic: ~2 characters per token for embedding models
+    return len(text) // 2
+
+
+def estimate_max_string_length(max_tokens: int) -> int:
+    """
+    Estimate maximum string length based on token limit.
+
+    Args:
+        max_tokens: Maximum token limit
+
+    Returns:
+        Estimated maximum string length
+    """
+    # Conservative estimation: 1.8 characters per token for embedding models
+    return int(max_tokens * 1.8)
+
+
+def truncate_text_gradually(text: str, max_tokens: int, steps: int = 5) -> str:
+    """
+    Gradually truncate text using step-wise approach.
+
+    Args:
+        text: Input text to truncate
+        max_tokens: Maximum token limit
+        steps: Number of truncation steps to try
+
+    Returns:
+        Truncated text that should be within token limits
+    """
+    if not text:
+        return text
+
+    estimated_tokens = estimate_token_count(text)
+    if estimated_tokens <= max_tokens:
+        return text
+
+    # Log warning about truncation
+    estimated_max_length = estimate_max_string_length(max_tokens)
+    logger.warning(
+        f"Text too long for embedding model: "
+        f"estimated {estimated_tokens} tokens > {max_tokens} limit. "
+        f"Estimated max string length: {estimated_max_length} chars. "
+        f"Will truncate gradually in {steps} steps."
+    )
+
+    # Calculate truncation steps
+    reduction_per_step = 0.8  # Reduce by 20% each step
+
+    current_text = text
```

Comment:
Will this work?
<img width="827" height="176" alt="image" src="https://github.com/user-attachments/assets/7ba6de04-11b6-4262-ab4a-8a3bc7017d5f" />

Reply:
<img width="810" height="883" alt="image" src="https://github.com/user-attachments/assets/99ddf921-51d7-4561-98f4-1a0c10d60618" />